### DISCUSSION TITLE
Release mode app freeze

### DISCUSSION THREAD

**Discussion by Ghostbird** [dotnet/maui#13472] [Q&A] [answered]
I'm asking this as a discussion item, not as an issue, since I cannot reproduce this in any minimal example, and I've no clue what the actual underlying issue is.

I've got a MAUI app that works fine in Debug mode. However the Release mode build will start-up and just completely get locked up somehow. However it only happens about 80%-90% of the time. Sometimes it will work fine.

Visually, the app starts, and the first page opens, the loading spinner is shown, and then the loading spinner stays on the screen, and the app doesn't react to any input, such as the button to open the in-app settings page. However the spinner keeps spinning smoothly.

This app has extensive debug logging to Android's logcat, but there's nothing in the logs that even hints at something going wrong. Only if I start trying to use the UI, Android will pop-up an _app unresponsive_ dialogue about 30 seconds later.

I've analysed as much of the code as I could for deadlocks, but I haven't found anything yet. Does anyone have a good idea on how to tackle this?

**Comment by JohnHDev** [comment, marked as answer]
I would suggest commenting out the code that is behind the spinning dialog, I presume its an async event to fetch data that then syncs with the UI? Don't call it, see if that app is then responsive. Keep commenting out until the app is responsive, and then start adding back code until you identify the issue.
Its a little bit old school, but Ive had to use this in the past when unable to debug an issue directly.

**Comment by Ghostbird** [comment]
Ah, found it 🤦🏽 There's _one_ explicit synchronisation context in the app. That's the one that's deadlocking. How did I miss that! Quite weird that it _only_ does that in release mode though, when I stepped through it in debug mode, it worked perfectly.

**Reply by Ghostbird** [reply]
I wrapped the whole construct in a `try { …original after semaphore acquisition block… } finally { semaphore.Release(); }` construct, and now it works fine in release mode too. My guess is that the optimisation rewrote the piece in an unexpected manner, that is prevented by this construct.
